python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -119,6 +121,75 @@ def categorical_embedding_with_indices(feature_tensor, feature_info, file_io: Fi
     return embedding
 
 
+def categorical_embedding_to_encoding_bilstm(feature_tensor, feature_info, file_io: FileIO):
+    """
+    Encode a string tensor into categorical embedding.
+    Works by converting the string into a word sequence and then generating a categorical/char embedding for each words
+    based on the List of strings that form the vocabulary set of categorical values, defined by the argument
+    vocabulary_file.
+    The char/byte embeddings are then combined using a biLSTM.
+
+    Args:
+        feature_tensor: String feature tensor that is to be encoded
+        feature_info: Dictionary representing the feature_config for the input feature
+
+    Returns:
+        Encoded feature tensor
+
+    Args under feature_layer_info:
+        vocabulary_file: string; path to vocabulary CSV file for the input tensor containing the vocabulary to look-up.
+                        uses the "key" named column as vocabulary of the 1st column if no "key" column present.
+        max_length: int; max number of rows to consider from the vocabulary file.
+                        if null, considers the entire file vocabulary.
+        embedding_size: int; dimension size of the embedding;
+                        if null, then the tensor is just converted to its one-hot representation
+        encoding_size: int: dimension size of the sequence encoding computed using a biLSTM
+
+    NOTE:
+        The input dimension for the embedding is fixed to 256 because the string is
+        converted into a bytes sequence.
+    """
+    args = feature_info.get("feature_layer_info")["args"]
+
+    vocabulary_df = file_io.read_df(args["vocabulary_file"])
+    if "key" in vocabulary_df.columns:
+        vocabulary_keys = vocabulary_df["key"]
+    else:
+        vocabulary_keys = vocabulary_df.iloc[:,0]
+    if "max_length" in feature_info:
+        vocabulary_keys = vocabulary_keys[:feature_info["max_length"]]
+    vocabulary_keys = vocabulary_keys.fillna(feature_info["default_value"]).values
+
+    vocabulary_ids = (
+        vocabulary_df["id"].values if "id" in vocabulary_df else list(range(len(vocabulary_keys)))
+    )
+
+    num_oov_buckets = args.get("num_oov_buckets", 1)
```

Comment:
num_oov_buckets is required later in the stage.
Does it make sens to have num_oov=1 for a label? I think we should make it optional (especially for the label in classification).

Reply:
You can set it to 0 in the feature config, no? But to keep it from breaking, would be useful to set to 1 nonetheless. Up to you.

Reply:
labels don't have an embedding layer, do they?  No need for oov?

Reply:
Yep. Why are you passing label through an embedding layer? @Ullimague

Reply:
I can put default 0 here, but not from the config, because from the config we end-up having:
```
    if num_oov_buckets <= 0:
      raise ValueError("oov_buckets must be > 0.")
```
from tensorflow StaticVocabularyTable
called [here](https://github.com/salesforce/ml4ir/blob/37540c26c01585264f125f9eaeb4db06c44d7486/python/ml4ir/base/features/feature_fns/categorical.py#L156)

Reply:
This comment is for line 438. I mixed the 2 functions because it didn't appeared as modified in the review..